Add support for Windows

EnzoMartin opened the pull request:
- Hides folders instead of coloring them
- Adds a .txt file in `node_modules` to remind the user they've hidden directories with john
- Makes `finder-tag` module optional
- Add optional `fswin` module

Need to make the tests run as well, and update the readme, but wanted to get input on it before doing that.

Since Windows doesn't support Tags, I thought a decent alternative would be to hide the folders. It does cause some slight differences with the options though since you need to specify whether or not to hide the dependencies.

davej commented:
Cool, I think this makes sense for windows. :+1: 

One thing worth doing is adding a `lib` folder with the OS specific implementations (`win32.js` and `darwin.js`), that will clean things up a bit.
In `index.js` we can still do the platform agnostic stuff and import the os-specific stuff using something like:
```js
const lib = require('./lib/' + process.platform);
```

Does that make sense?

EnzoMartin commented:
Yep makes sense, will add additional commits and split that up nicely tomorrow

Not sure why the tests are failing though, I don't have a Mac on hand to verify but my changes shouldn't be a breaking change for OSX

davej commented:
Yeah the logs aren't being shown on travis because of some issues https://www.traviscistatus.com/

I'll pull down the branch and test it locally.

davej commented:
So tests run fine it's just linting errors from [xo](https://www.npmjs.com/package/xo):
```
index.js
   11:53  error  Missing space before function parentheses             space-before-function-paren
   11:55  error  Missing space before opening brace                    space-before-blocks
   12:78  error  Missing space before function parentheses             space-before-function-paren
   12:80  error  Missing space before opening brace                    space-before-blocks
   63:12  error  Missing semicolon                                     semi
   86:11  error  Keyword "if" must be followed by whitespace           space-after-keywords
   86:18  error  Missing space before opening brace                    space-before-blocks
  100:15  error  Keyword "if" must be followed by whitespace           space-after-keywords
  100:39  error  Missing space before opening brace                    space-before-blocks
  104:15  error  Keyword "if" must be followed by whitespace           space-after-keywords
  104:42  error  Missing space before opening brace                    space-before-blocks
  117:15  error  Expected { after 'if' condition                       curly
  123:24  error  Missing space before value for key "dependencies"     key-spacing
  124:27  error  Missing space before value for key "devDependencies"  key-spacing
```

Just running npm test should work on windows for linting because linting is done before the tests are run.

davej commented:
By the way, travis is set to run the tests on OS X at the moment. I'll try and figure out if it's possible to do tests on both OS X and Windows. I'm sure it's possible, I've just never done it before.

EnzoMartin commented:
Ah, sorry about that, will fix those as well then

davej commented:
I've made some refactoring in preparation for windows support so pull down the latest on master.

Let me know if anything is unclear. I believe that you'll only need to implement the `handleOptions` and `performAction` methods.

EnzoMartin commented:
Refactored to use the `win32` file as discussed, I also slightly modified the `performAction` method so it's fully contained within each platform specific `lib` file, let me know if you're ok with that.

I added 2 TODOs as well which I'd like to add, one is for simplifying the supported platform since that information is already in the `package.json` and the other is to support the `optionalDependencies`. I'll add support for both later today.

Will also add the tests for Windows later today, heading in to the office now

davej commented on index.js:
```diff
@@ -27,28 +27,16 @@ module.exports = (projectPath, opts) => {
     }
 
-    opts = Object.assign({}, defaultOpts, opts || {});
+    opts = Object.assign({}, defaultOpts, opts || {}, {projectPath});
```
What is this for?

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,62 @@
+  if (opts.clear) {
+    actions.push(showModules(opts.projectPath, [], true));
+  } else {
+    // TODO: Support optional dependencies
```
Optional dependencies should already be in your dependencies object, because of normalization from [`read-pkg-up`](https://github.com/sindresorhus/read-pkg-up).

EnzoMartin replied:
Ah, nice, hadn't checked, will remove that TODO then :+1:

davej commented:
I still think we can do `handleDeps` in a generalised way. We're doing a lot of the same things just slightly different implementations. I've added a `performPreAction` method for you to implement in `win32`, I think that's all you need.

`win32.performPreAction` should:
* hide all `node_modules`
* create text file saying there is hidden files
* return promise when done

`win32.performAction` should:
* show `node_modules` from `deps`
* return promise when done
  * promise should contain a signature like:

```js
[
  {
    path: '/path/to/project/node_modules/finder-tag',
    hidden: true,
    module: 'finder-tag'
  },
  {…},
  {…}
]
```

I can give a bit more detail if anything isn't clear. Hope that all makes sense :-)

EnzoMartin commented:
Ahh alright, yep, will do

davej commented:
FYI: [Refactored the cli stuff](https://github.com/davej/john/commit/5bc02f5bff0fa4bf115f04334cc989343ba9ff61) for easier multi-platform support.

EnzoMartin commented:
Alrighty. So.

Update all the methods to use the same as the OSX ones, I added a rule override to disable the checking of linebreak styles since Windows uses CRLF vs LF and it doesn't have any impact.

I also fixed a bug that would prevented the `getDepInfo` promise from settling if you had more dependencies declared than you had installed (such as failed optional dependencies.

davej commented:
Awesome, thanks for sticking with it. Looks good at first glance, I'll go through it properly this evening. :+1:

EnzoMartin commented:
And tests added :+1:

EnzoMartin commented:
Alright, updated readme and various other bits and bobs, should be all set now

EnzoMartin commented on package.json:
```diff
@@ -20,6 +27,7 @@
     "test": "xo && ava"
   },
   "files": [
+    "lib",
```
Without this, the `lib` folder will be missing when doing `npm install`

davej replied:
Good catch, I overlooked this.

davej commented on index.js:
```diff
@@ -29,8 +30,10 @@ module.exports = (projectPath, opts) => {
   const getDepInfo = deps =>
     Object.keys(deps || {}).map(dep => {
       const depPath = path.join(projectPath, 'node_modules', dep);
+      // Catch handles dependencies that are not installed
```
Good find!

EnzoMartin replied:
Hehe, spent an hour trying to figure out why everything was failing silently. Promises are neat, but can be a real pain in the ass to debug and figure out where in the chain it's failing.

I was debating adding `catch` on every promise but figured that's better left for after this PR is in

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,152 @@
+    {flags: 'w'},
+    function (err) {
+      if (err) {
+        console.warn('Failed to write info file', err);
```
I prefer to only output to console in the CLI, console output isn't very useful when used programmatically. Perhaps we should just return an error? If we can't write a file to the `node_modules` folder then something is probably very wrong and it will fail when setting the hidden attribute anyway.

Also for consistency perhaps we can use `pify` on `fs.writeFile` and do a promise instead?

EnzoMartin replied:
Sure thing

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,152 @@
+        // Hide all the directories
+        Promise.all(files.map(
+          (file) => new Promise((resolve) =>
+            fswin.setAttributes(file, {IS_HIDDEN: hidden}, () => resolve(file))
```
Should this promise reject if `setAttributes` fails?

EnzoMartin replied:
Sure, no problem

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,152 @@
+
+  return Promise.all(deps.map(dep =>
+    new Promise((resolve) =>
+      fswin.setAttributes(dep.path, {IS_HIDDEN: hide}, () => resolve(Object.assign(dep, {hidden: hide})))
```
Should this promise reject if setAttributes fails?

EnzoMartin replied:
Sure, no problem

davej commented on package.json:
```diff
@@ -36,22 +44,30 @@
     "pify": "^2.3.0",
     "read-pkg-up": "^1.0.1"
   },
+  "optionalDependencies": {
```
Pity npm doesn't give a way to specify OS-specific dependencies. I'm not sure `optionalDependencies` is what we want because they will install anyway but if they fail for a legitimate reason during installation then it will be ignored. What do you think? :-/

EnzoMartin replied:
Yup correct

davej commented on readme.md:
```diff
@@ -49,6 +67,28 @@ $ john --help
+  Options
+    --clear     Clear all hidden dependencies. [Default: false]
+    --deps      Hide dependencies. [Default: false]
+    --dev-deps  Hide devDependencies. [Default: false]
```
I'm a bit unclear on the last two options. Is the `default` supposed to be `true` for `deps` and `dev-deps`?

EnzoMartin replied:
Yea, I was debating that too, I figured the brevity is more desirable

davej commented on readme.md:
```diff
@@ -1,28 +1,46 @@
+#### On Windows
+He hides away non top-level dependencies and devDependencies, leaving you with just the modules that are important to you.
+
+<p align="center"><img src ="https://zippy.gfycat.com/ShimmeringThirstyAmericanwarmblood.gif" /></p>
```
Awesome, thanks :)

davej commented:
Review finished. Everything is great, just a few very minor comments and then it's good for merge.

EnzoMartin commented:
Alrighty, made the changes outlined above

EnzoMartin commented:
As a side note, the reason `setAttributes` isn't wrapped with `pify` is because it was randomly silently failing with it, I'm not sure why though and after investigating for a while decided it wasn't really worth the effort in tracking it down as the slightly more verbose promise works fine.

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,156 @@
+  const modules = path.join(projectPath, 'node_modules');
+  const hidden = typeof opts.clear === 'boolean' ? !opts.clear : true;
+
+  writeFile(
```
I don't think that will return an error, looks like it will be silently swallowed up. You could do something like:

```js
const writeTextFile = writeFile(
  path.join(modules, 'modules_hidden.txt'),
  'You\'ve hidden modules with `john` - https://github.com/davej/john',
  {flags: 'w'}).catch((err) => new Error('Failed to write info file', err)
);

return writeTextFile.then(
  readdir(modules)
    .then(...)
)
```

EnzoMartin replied:
Mm, well, the file thing is just a nicety I felt, so I'm not sure it's really worth caring about it, much less making it block the rest of the script, thoughts?

davej replied:
Ok, well if you don't want to throw an error then I guess the original `console.warn` is the best approach.

EnzoMartin replied:
Actually, that still won't throw the error as there's no catch higher up

EnzoMartin replied:
```javascript
const writeTextFile = writeFile(
    path.join(modules, 'modules_hidden.txt'),
    'You\'ve hidden modules with `john` - https://github.com/davej/john',
    {flags: 'w'}
  ).catch(
    (err) => {
      throw new Error('Failed to write info file', err);
    }
  );
```

Will throw the error though

davej replied:
A rejected promise returned in `performPreAction` will propagate all the way up to the promise that is returned in `index.js`.

davej replied:
Ah I understand, yeah the promise wont reject unless it's thrown (or `Promise.reject()`).

EnzoMartin replied:
Right.. I think this is a sign I need to go to bed and pick it up in the morning :P

davej replied:
You could just remove the `catch` statement entirely though, node should give you a descriptive enough error.

EnzoMartin replied:
> Ah I understand, yeah the promise wont reject unless it's thrown (or Promise.reject()).

Aha, so I'm not crazy. Yea I did some real dirty quick testing with it, it's either wrap it and call `.reject` or invoke throw.

Letting it go through catch as is will allow the execution to continue but also output the error message

davej replied:
To be honest, I think the best solution is just to remove the `catch` function. Node should give you a descriptive error anyway. :)

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,156 @@
+    {flags: 'w'}).catch((err) => new Error('Failed to write info file', err)
+  );
+
+  return Promise.all([
```
I'm not sure the `Promise.all()` is actually needed here, I think you could just do `return readdir(modules).then(…)`. Either way, it won't cause any issues though.

EnzoMartin replied:
I'll remove it, it's not needed indeed

davej commented on lib/win32.js:
```diff
@@ -0,0 +1,156 @@
+        files = files.filter((file) => file);
+
+        // Hide all the directories
+        Promise.all(files.map(
```
You forgot to return this promise. :)

EnzoMartin replied:
Doh

davej commented:
Nearly there now, see comments!

I setup CI for Windows. Looks like the windows build is timing out on one of the tests, can you confirm that the tests are running ok on your machine? https://ci.appveyor.com/project/davej/john/build/1.0.4/job/lgkd31bcb415aiqo

> As a side note, the reason setAttributes isn't wrapped with pify is because it was randomly silently failing with it.

Yeah looks like that is because the `fswin.setAttributes` API doesn't follow the node.js convention of `(err, callback) => {}`.

EnzoMartin commented:
Yea looks like the clear and show test is not returning/completing/running

davej commented:
Not sure if it's related to the failing tests but something to be aware of with ava is that the tests are run in parallel. It's possible there could be a write lock or some other kind of race condition in the tests. If you want a test to run serially then use `test.serial()` instead of just `test()`.

EnzoMartin commented:
There's a failure happening somewhere inside the preAction, debugging..

EnzoMartin commented:
I've decuded that that `fswin` library can burn in hell.

davej commented:
Haha, did you try changing the tests to `test.serial()` and keeping `fswin` async.

EnzoMartin commented:
Yea, tried that, I removed all but one test as well, still failed. It *seems* that something is eating one or more of the `setAttributes` promises and they simply never settle, so preAction never settles and the whole thing becomes stuck.

I took the liberty of upgrading `ava` to `0.11` as well, fixes the deprecation spam and a few other things

davej commented:
Awesome. Builds are passing. Perhaps, [squash your commits](http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit) if you don't mind and then you're a free man! :)

EnzoMartin commented:
What's more peculiar is that none of the `setAttribute` calls return with `success === false` or throw an error, so I think it's somewhere inside the library that it's going wonky and entering limbo. Was veryyyyy close to throwing the whole library out and just writing the bit I need for setting the hidden flag.

EnzoMartin commented:
Do you want me to squash everything into 1 commit?

davej commented:
Whatever you like really. If there's a couple of logical bits that make sense then you can squash it to a few commits. It's really just to get rid of all the merges from the git history.

EnzoMartin commented:
Alright, not feeling up to the volatile black voodoo magic that that involves, will tackle it in the morning :+1:

davej commented:
Cool, even rockstars gotta sleep!

davej commented:
FYI: Curiosity got the better of me and I booted up a windows VM to do some testing. You're correct that it's basically a `fswin` bug, I was able to get it working by adding a `setTimeout` around the `fswin` call.

John will probably be mostly used on the command line so it's not essential that it has to be async. I'll leave it up to you whether to use the sync method or use the `setTimeout` hack. Whichever you choose, use it in both calls to `setAttributes` though, because the other call will fail eventually if there are enough top-level dependencies in the target project.

Here is the [`setTimeout` version of win32.js](https://gist.github.com/davej/9b76328dcdb6cea87a24).

EnzoMartin commented:
Hah, nice. We should open an issue in his library then, rather than add a workaround, the setTimeout is a bit hacky IMO

davej commented:
Yup, happy enough to merge with the synchronous method for the time being though.

EnzoMartin commented:
Alrighty, I *think* I squashed everything correctly

davej commented:
Can you make both `setAttributes` calls synchronous, the other call will most likely fail too if there are more top-level dependencies.

davej commented:
No need to re-squash.

EnzoMartin commented:
Oh, sorry about that, thought I got both, hah

EnzoMartin commented:
Alright, there we go.

I redid the commit as I forgot to mention the 2 bugs fixed with it as well

davej commented:
Whoop, the epic saga of PR#1 is complete. Congrats! :-)

I'll publish a new version later today.

EnzoMartin commented:
:tada: \o/ :tada: 

Should update the description on GitHub as well
